Firestore/core/src/firebase/firestore/util/async_queue.h

```diff
@@ -0,0 +1,204 @@
+  explicit DelayedOperation(
+      const std::shared_ptr<detail::DelayedOperationImpl>& instance)
+      : handle_{instance} {
+  }
+
+  std::weak_ptr<detail::DelayedOperationImpl> handle_;
+};
+
+class AsyncQueue {
+ public:
+  using Milliseconds = std::chrono::milliseconds;
+  using Operation = std::function<void()>;
+
+  explicit AsyncQueue(const dispatch_queue_t dispatch_queue)
+      : dispatch_queue_{dispatch_queue} {
+  }
+
+  /**
+   * Asserts that we are already running on this queue (actually, we can only
+   * verify that the queue's label is the same, but hopefully that's good
+   * enough).
+   */
+  void VerifyIsCurrentQueue() const;
+
+  /**
+   * Declares that we are already executing on the correct `dispatch_queue_t`
+   * and would like to officially execute code on behalf of this `AsyncQueue`.
+   * To be used only when called back by some other API directly onto our queue.
+   * This allows us to safely dispatch directly onto the worker queue without
+   * destroying the invariants this class helps us maintain.
+   */
+  void EnterCheckedOperation(const Operation& operation);
+
+  /**
+   * Same as `dispatch_async()` except it asserts that we're not already on the
+   * queue, since this generally indicates a bug (and can lead to re-ordering of
+   * operations, etc).
+   *
+   * @param operation The operation to run.
+   */
+  void Enqueue(const Operation& operation);
+
+  /**
+   * Unlike `Enqueue`, this method does not require you to dispatch to a
+   * different queue than the current one (thus it is equivalent to a raw
+   * `dispatch_async()`).
+   *
+   * This is useful, e.g. for dispatching to the user's queue directly from user
+   * API call (in which case we don't know if we're already on the user's queue
+   * or not).
+   *
+   * @param operation The operation to run.
+   */
+  void EnqueueAllowingSameQueue(const Operation& operation);
+
+  /**
+   * Schedules an operation after the specified delay.
+   *
+   * Unlike `Enqueue`, this method does not require you to dispatch to a
+   * different queue than the current one.
+   *
+   * The returned `DelayedOperation` handle can be used to cancel the operation
+   * prior to its running.
+   *
+   * @param operation The operation to run.
+   * @param delay The delay after which to run the operation.
+   * @param timer_id A `TimerId` that can be used from tests to check for the
+   *     presence of this operation or to schedule it to run early.
+   * @return A `DelayedOperation` instance that can be used for cancellation.
+   */
+  DelayedOperation EnqueueAfterDelay(Milliseconds delay,
+                                     TimerId timer_id,
+                                     Operation operation);
+
+  /**
+   * Wrapper for `dispatch_sync()`. Mostly meant for use in tests.
+   *
+   * @param operation The operation to run.
+   */
+  void RunSync(const Operation& operation);
+
+  /**
+   * For tests: determine if a delayed operation with a particular `TimerId`
```

Comment:
Just my 1 cent, those helper methods purely for tests can be defined `private` and make test friend class rather than expose them `public`. From maintenance point of view, change private methods is easier that changing a public one, the latter of which may involve larger scope of callers.

Reply:
Mental exercise: It might be neat to create a 'test' namespace nested in this one, with a 'DelayedOperationTester' class. You could make these methods private here, and mirror them as public in the Tester. It'd make it more obvious that they exist only for testing purposes and shouldn't be otherwise relied on.

Optional.

Reply:
I completely agree with the general idea, but I don't think I will have enough context to figure out what the best solution is until I get to changing Objective-C code to using this version of queue.

Another thing to note: the vector of `DelayedOperation`s exists mostly for the purpose of tests. For cancellation, just transferring ownership of `DelayedOperation` libdispatch should be enough.